components/supervisor/pkg/ports/tunneled-ports.go

```diff
@@ -0,0 +1,296 @@
+// Copyright (c) 2021 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package ports
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net"
+	"sort"
+	"strconv"
+	"sync"
+
+	"github.com/gitpod-io/gitpod/supervisor/api"
+	"golang.org/x/xerrors"
+)
+
+type PortTunnelDescription struct {
+	LocalPort  uint32
+	TargetPort uint32
+	Visibility api.TunnelVisiblity
+}
+
+type PortTunnelState struct {
+	Desc    PortTunnelDescription
+	Clients map[string]uint32
+}
+
+type PortTunnel struct {
+	State PortTunnelState
+	Conns map[string]map[net.Conn]struct{}
+}
+
+type TunnelOptions struct {
+	SkipIfExists bool
+}
+
+// TunneledPortsInterface observes the tunneled ports.
+type TunneledPortsInterface interface {
+	// Observe starts observing the tunneled ports until the context is canceled.
+	// The list of tunneled ports is always the complete picture, i.e. if a single port changes,
+	// the whole list is returned.
+	// When the observer stops operating (because the context as canceled or an irrecoverable
+	// error occured), the observer will close both channels.
+	Observe(ctx context.Context) (<-chan []PortTunnelState, <-chan error)
+
+	// Tunnel opens new tunnels.
+	Tunnel(ctx context.Context, options *TunnelOptions, descs ...*PortTunnelDescription) ([]uint32, error)
+
+	// CloseTunnel closes tunnels.
+	CloseTunnel(ctx context.Context, localPorts ...uint32) ([]uint32, error)
+
+	// EstablishTunnel actually establishes the tunnel
+	EstablishTunnel(ctx context.Context, clientID string, localPort uint32, targetPort uint32) (net.Conn, error)
+}
+
+// TunneledPortsService observes the tunneled ports.
+type TunneledPortsService struct {
+	tunnelsMu sync.RWMutex
+	tunnels   map[uint32]*PortTunnel
+
+	changesMu sync.RWMutex
+	changes   map[chan struct{}]struct{}
```

Comment:
better use a `sync.Cond` that re-uses the `tunnelsMu` as `sync.Locker`. This way we don't need to faff with maps of chans or actively maintain subscriptions.